File: scripts/openapi.ts
#!/usr/bin/env bun

import { mkdir } from "node:fs/promises";
import { dirname } from "node:path";
import { parseArgs, runCli } from "./lib/cli";
import { Repo } from "./lib/repo";

const USAGE = "usage: scripts/openapi.ts generate | check";

const SPEC_PATH = "crates/omnifs-api/openapi/daemon.json";

await runCli(async () => {
  const { positionals } = parseArgs(Bun.argv.slice(2), {});
  const [command, ...rest] = positionals;
  if (rest.length > 0) throw new Error(`unexpected argument: ${rest[0]}`);

  const workspace = new OpenApiWorkspace(await Repo.discover());
  if (command === "generate") {
    await workspace.generate();
    return;
  }
  if (command === "check") {
    await workspace.check();
    return;
  }
  throw new Error(USAGE);
});

class OpenApiWorkspace {
  constructor(private readonly repo: Repo) {}

  async generate(): Promise<void> {
    await this.writeFileIfChanged(SPEC_PATH, await this.specFromImplementation());
    console.log("OpenAPI spec written");
  }

  async check(): Promise<void> {
    const implemented = await this.specFromImplementation();
    const checkedIn = await Bun.file(this.repo.path(SPEC_PATH)).text();
    if (checkedIn !== implemented) {
      throw new Error(`checked-in OpenAPI spec is stale; run \`just openapi\``);
    }
    console.log("OpenAPI spec is current");
  }

  private async specFromImplementation(): Promise<string> {
    return await this.repo.$`cargo run -p omnifs-daemon --bin omnifsd-openapi --quiet`.text();
  }

  private async writeFileIfChanged(path: string, contents: string): Promise<void> {
    const file = Bun.file(this.repo.path(path));
    if (await file.exists() && await file.text() === contents) return;
    await mkdir(dirname(this.repo.path(path)), { recursive: true });
    await Bun.write(this.repo.path(path), contents);
  }
}

File: crates/xtask/src/openapi.rs
use std::fs;
use std::path::Path;
use std::process::Command;

use anyhow::{Context, Result, bail};

const SPEC_PATH: &str = "crates/omnifs-api/openapi/daemon.json";

/// Regenerate the checked-in OpenAPI spec from the daemon implementation.
pub fn generate(root: &Path) -> Result<()> {
    let spec = spec_from_implementation(root)?;
    let path = root.join(SPEC_PATH);
    fs::write(&path, spec).with_context(|| format!("write {}", path.display()))?;
    println!("OpenAPI spec written");
    Ok(())
}

/// Fail if the checked-in OpenAPI spec differs from the daemon implementation.
pub fn check(root: &Path) -> Result<()> {
    let implemented = spec_from_implementation(root)?;
    let checked_in =
        fs::read_to_string(root.join(SPEC_PATH)).with_context(|| format!("read {SPEC_PATH}"))?;
    if checked_in != implemented {
        bail!("checked-in OpenAPI spec is stale; run `just openapi`");
    }
    println!("OpenAPI spec is current");
    Ok(())
}

fn spec_from_implementation(root: &Path) -> Result<String> {
    let output = Command::new("cargo")
        .args([
            "run",
            "-p",
            "omnifs-daemon",
            "--bin",
            "omnifsd-openapi",
            "--quiet",
        ])
        .current_dir(root)
        .output()
        .context("run omnifsd-openapi")?;
    if !output.status.success() {
        bail!(
            "omnifsd-openapi failed: {}",
            String::from_utf8_lossy(&output.stderr).trim()
        );
    }
    String::from_utf8(output.stdout).context("omnifsd-openapi produced non-utf8 output")
}
